lib/dockerfile.ts

```diff
@@ -0,0 +1,223 @@
+	 */
+	fileDependencies: FileDependency[];
+	/**
+	 * The directory in the container which these commands should be
+	 * run in
+	 */
+	workDir: string;
+}
+
+class Dockerfile {
+
+	public actionGroups: DockerfileActionGroup[] = [];
+
+	public constructor(dockerfileContents: string) {
+		this.parse(dockerfileContents);
+	}
+
+	private parse(dockerfileContents: string) {
+		const entries = parser.parse(dockerfileContents, {
+			includeComments: false,
+		});
+
+		let workDir = '/';
+		let commands: Command[] = [];
+		let lastCopy: FileDependency[] = [];
+
+
+		_.each(entries, (entry) => {
+
+			switch (entry.name.toUpperCase()) {
+				case 'ADD':
+					// This isn't supported, as adding urls etc can really mess things up.
+					throw new UnsupportedError('Dockerfiles containing the ADD instruction are not supported. Please use COPY.');
+				case 'COPY':
+					// If there is any commands which have occured already, save them as an action
+					// group
+					if (commands.length > 0) {
+						this.actionGroups.push({
+							workDir,
+							commands,
+							fileDependencies: lastCopy,
+						});
+						commands = [];
+					}
+					if (!_.isArray(entry.args)) {
+						throw new DockerfileParseError('Non-array arguments passed to COPY');
+					}
+
+					lastCopy = Dockerfile.copyArgsToFileDeps(workDir, entry.args);
+					break;
+				case 'WORKDIR':
+					// Take every command we currently have, and save it as an action group,
+					// as encountering another working directory means that every command
+					// following must execute in a different action group
+					if (commands.length > 0) {
+						this.actionGroups.push({
+							workDir,
+							commands,
+							fileDependencies: lastCopy,
+						});
+					}
+					commands = [];
+					if (!_.isString(entry.args)) {
+						throw new DockerfileParseError('Non-string argument to WORKDIR');
+					}
+					workDir = entry.args;
+					break;
+				case 'RUN':
+					// Add this to the current set of commands
+					commands.push(Dockerfile.processRunArgs(entry.args));
+					break;
+			}
+		});
+
+		// Store any commands left over
+		if (commands.length > 0) {
+			this.actionGroups.push({
+				workDir,
+				commands,
+				fileDependencies: lastCopy,
+			});
+		}
+	}
+
+	// TODO: We can probably cache this value
+	/**
+	 * Get a list of action groups which should be re-executed if any of
+	 * the files passed to this function have changed.
+	 *
+	 * @param files A list of files whose contents have changed. The path
+	 * should be relative to the build context
+	 */
+	public getActionGroupsFromChangedFiles(files: string[]): DockerfileActionGroup[] {
+		let actionGroupIdx = 0;
+		for (const actionGroup of this.actionGroups) {
+			// FIXME: COPY src/ src/ won't work due to the fact that there isn't a glob on the end of
+			// the target - we can fix this by silenty converting src/ to src/*
+
+			// For every dependency, see if it matches the current action group
+			const matches = _.uniq(_.reduce(
```

Comment:
This can be
```typescript
_(actionGroup.fileDependencies).flatMap({{ localPath }) => {
	return minimatch.match(files, localPath);
}).uniq().value()